File: 3433. Count Mentions Per User.cpp
class Solution {
public:
    vector<int> countMentions(int numberOfUsers, vector<vector<string>>& events) {

        sort(events.begin(), events.end(),
             [](const vector<string>& a, const vector<string>& b) {
                 int t1 = stoi(a[1]);
                 int t2 = stoi(b[1]);
                 if (t1 != t2) return t1 < t2;
                 return (a[0] == "OFFLINE") ? true : false;
             });

        vector<int> mentions(numberOfUsers, 0);
        unordered_set<int> onlineUsers;
        vector<int> offlineUsers(numberOfUsers, 0);

        for (int i = 0; i < numberOfUsers; i++) {
            onlineUsers.insert(i);
        }

        for (auto& event : events) {
            int currTime = stoi(event[1]);

            for (int i = 0; i < numberOfUsers; i++) {
                if (!onlineUsers.count(i) && offlineUsers[i] + 60 <= currTime)
                    onlineUsers.insert(i);
            }

            if (event[0] == "MESSAGE") {

                if (event[2] == "ALL") {
                    for (int i = 0; i < numberOfUsers; i++) mentions[i]++;
                }

                else if (event[2] == "HERE") {
                    for (int i = 0; i < numberOfUsers; i++)
                        if (onlineUsers.count(i)) mentions[i]++;
                }

                else {
                    string str = event[2];
                    int curr = 0;

                    for (char c : str) {
                        if (c >= '0' && c <= '9') {
                            curr = curr * 10 + (c - '0');
                        } 
                        else if (c == ' ') {
                            mentions[curr]++;
                            curr = 0;
                        }
                    }

                    mentions[curr]++; // last user
                }
            }

            else { // OFFLINE
                int id = stoi(event[2]);
                onlineUsers.erase(id);
                offlineUsers[id] = stoi(event[1]);
            }
        }

        return mentions;
    }
};

File: 3433. Count Mentions Per User.java
class Solution {
    public int[] countMentions(int numberOfUsers, List<List<String>> events) {
        Collections.sort(events, (a, b) -> Integer.parseInt(a.get(1)) != Integer.parseInt(b.get(1)) ? Integer.compare(Integer.parseInt(a.get(1)), Integer.parseInt(b.get(1))) : a.get(0).equals("OFFLINE") ? -1 : 1);
        int[] mentions = new int[numberOfUsers];
        HashSet<Integer> onlineUsers = new HashSet<>();
        int[] offlineUsers = new int[numberOfUsers];
        for(int i = 0; i < numberOfUsers; i++) onlineUsers.add(i);

        for(List<String> event : events){
            int currTime = Integer.parseInt(event.get(1));
            for(int i = 0; i < numberOfUsers; i++){
                if(!onlineUsers.contains(i) && offlineUsers[i] + 60 <= currTime) onlineUsers.add(i);
            }
            if(event.get(0).equals("MESSAGE")){
                if(event.get(2).equals("ALL")){
                    for(int i = 0; i < numberOfUsers; i++) mentions[i]++;
                }
                else if(event.get(2).equals("HERE")){
                    for(int i = 0; i < numberOfUsers; i++) if(onlineUsers.contains(i)) mentions[i]++;
                }
                else{
                    String str = event.get(2);
                    int curr = 0;
                    for(int i = 0; i < str.length(); i++){
                        if(str.charAt(i) >= '0' && str.charAt(i) <= '9'){
                            curr *= 10;
                            curr += (str.charAt(i) - '0');
                        }
                        else if(str.charAt(i) == ' '){
                            mentions[curr]++;
                            curr = 0;
                        }
                    }

                    mentions[curr]++;
                }
            }
            else{
                int id = Integer.parseInt(event.get(2));
                onlineUsers.remove(id);
                offlineUsers[id] = Integer.parseInt(event.get(1));
            }
        }

        return mentions;
    }
}